### PR TITLE
Add option for parallax-aware zoom

### PR DESCRIPTION
**Pull request by Fireball-Teeworlds**
## How this works

Currently we have parallax X and Y parameters that control how far away a layer feels. Currently it's only taken into account for camera moving sideways.

This pull request makes zoom behave as if the camera was moving closer or further away, taking into account the distance to the layer.

In order to provide flexibility, this behaviour is using a separate parallax value, similar to how we have separate values for X and Y. Para Zoom will default to the minimum of parallax X, Y, clamped to be between 0 and 100. This seems to look well on most maps.

This pull request also introduces two new features to the Editor:
* Controls for configuring per-group Para Zoom value. If Custom Zoom is set to No, Para Zoom will automatically keep following to the default value described above, so that people new to the feature don't have to figure out how it works.
* Zoom button that previews the zoom behavior (next to Proof at the top).

### Para Zoom controls

screenshot_2022-08-22_21-38-04

screenshot_2022-08-22_21-37-58

### Zoom Button

screenshot_2022-08-22_21-40-46

screenshot_2022-08-22_21-40-50

## In-Game Screenshots

Video: https://youtu.be/W7eXQN0gRFI

This is an older version of the pull request that had an option to disable the new behaviour. The current version can only be disabled in map editor for a particular map.

### Springlobe 3

With new feature:
screenshot_2022-08-02_04-28-19

Without:
screenshot_2022-08-02_04-28-25

### Beyond Dreams

With new feature:
screenshot_2022-08-02_04-28-59

Without:
screenshot_2022-08-02_04-28-55

## Checklist

- [X] Tested the change ingame
- [X] Provided screenshots if it is a visual change
- [X] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [X] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by C0D3D3V** [comment]
Would this keep menus working if you zoom out, like that on Time Shop?

**Comment by def-** [comment]
Should we set this as default? Looks better than the old mode in your screenshots.

**Comment by C0D3D3V** [comment]
> Would this keep menus working if you zoom out, like that on Time Shop?

I just tested it, it does not make the situation better.

**Comment by Fireball-Teeworlds** [comment]
Some ideas on how it can be enabled.
### A: Client setting (this version but with default to enabled)
* Looks good on most maps. Doesn't break anything on maps that I've checked. Doesn't affect default zoom
* Does make some maps slightly less good looking in a certain zoom range. This can be addressed by tweaking parallax{x,y} values.

### B1: Add an option to set zoom-parallax value in the editor
(next to y-parallax and x-parallax. And if unset default to option A)
* Gives mappers more control
* Might be used to fix Time Shop (will need to check)
* Requires modifying map format (LayerGroup), unsure how much it will complicate the code/support - biggest consideration

### B2: Add option to disable at map level
* Can be skipped for maps where it doesn't work well

### C: Wait
The final option is to delay making the decision and instead gather more feedback by either temporarily enabling it as default or letting players enable it manually.

**Comment by def-** [comment]
So for B1 we would have different zoom settings per layer?
I guess B2 is easiest and we just need a new map config value, and can disable it for the maps where it doesn't work well.

**Comment by Fireball-Teeworlds** [comment]
> So for B1 we would have different zoom settings per layer?
It's similar to the existing parallax settings in that it configures the perceived distance from the camera and we have these per-layer

> I guess B2 is easiest and we just need a new map config value, and can disable it for the maps where it doesn't work well.
That's my understanding as well.

With B1 I'm mainly concerned about how complicated the map-reading/writing code will become.

**Comment by Fireball-Teeworlds** [comment]
Maybe do B2 and wait for feedback? We can get feedback from mappers if they notice that they want to tweak the zoom-parallax value independently from x/y parallax. And if not then the already known concern about it breaking maps will already be addressed.

**Comment by Chairn** [comment]
Given how it makes map rendering better, i'd say to get rid of the option and enforce it in the code. With the mapper setting, it should be enough to satisfy most people.

I still have to check with custom entities background and zoom between 5 and 10 which is what most player uses.

**Comment by Fireball-Teeworlds** [comment]
Looks like it's a bit tricky with map settings as well, as we don't have client-side settings that are reset on map change. Should I make it a "server setting" or add a mechanism for "map settings" (or game settings)?

**Comment by Fireball-Teeworlds** [comment]
Just confirmed that B1 will very likely be enough to fix Time Shop (will need to set Parallax Zoom to 0 for several shop layer groups).

**Comment by Fireball-Teeworlds** [comment]
Hm, looks like adding a new field isn't very difficult, we can also add some placeholder to Teeworlds to make sure they don't use the same field in future versions.

**Comment by Chairn** [comment]
Based on your last comment, it depends on layer?

Maybe use a per layer setting then but that would change map format.

**Comment by Fireball-Teeworlds** [comment]
Added per-group parameter. What do we think about the new field and the getter method?

I'm planning to add to the editor:
* A way to preview the effects of this setting in the editor, while allowing access to the edges of the layers when needed - probably with an off/on switch
* An option for Para Zoom parameter to be synced to maximum(para{x,y}) by default so that users unfamiliar with this feature can use the default good value instead of letting it out of sync unintentionally.

The second editor improvement might be important to add before merging this.

**Comment by Fireball-Teeworlds** [comment]
Pretty much ready. I need to fix the problem with Quad Env Points scaling (visual issue in the editor when Zoom button is enabled). But it's already usable and nothing seems broken with Zoom button switched off.

File: src/game/editor/editor.cpp [hunk 1]
```diff
@@ -865,6 +867,17 @@ void CEditor::DoToolbar(CUIRect ToolBar)
 
 		TB_Top.VSplitLeft(5.0f, nullptr, &TB_Top);
 
+		// zoom button
+		TB_Top.VSplitLeft(40.0f, &Button, &TB_Top);
+		static int s_ZoomButton = 0;
+		// TODO: do we need a keyboard shortcut?
```
Comment (Chairn) [inline src/game/editor/editor.cpp:873]: No need for keyboard shortcut because it's not something you do very often.

**Comment by def-** [comment]
Any concerns about the map version increment? That might conflict with Teeworlds doing an increase. Do we not care or should we try to get this change into Teeworlds too?

**Comment by Fireball-Teeworlds** [comment]
Maybe something like this? https://github.com/teeworlds/teeworlds/commit/580c01c2abeab094b573fab3bee882dee2e333df

**Comment by Fireball-Teeworlds** [comment]
TODO:
1. Run newest build with ASan & UBSan. Try to test unusual (<0, >100) Parallax Zoom values
2. Record a video with Time Shop fix process where new Editor controls are visible

**Comment by heinrich5991** [comment]
> Any concerns about the map version increment?

Yes, it should go into our own group object instead.

**Comment by Fireball-Teeworlds** [comment]
> Yes, it should go into our own group object instead.

Do you mean creating a new MapItem object and storing a reference (datafile index) to that object in the original one? (In order to make it easier to add fields in the future).

One benefit of using the existing object is that if teeworlds at some point implements zoom (either for spectating or as a gameplay change), we could reuse the same field as the field's function is not ddnet-specific.

**Comment by heinrich5991** [comment]
> Do you mean creating a new MapItem object

Yes.

> and storing a reference (datafile index) to that object in the original one? (In order to make it easier to add fields in the future).

No, I wouldn't modify the original object in any way. I'd maybe match them by their datafile item IDs (i.e. same ID of group and extended group objects describe the same group).

> One benefit of using the existing object is that if teeworlds at some point implements zoom (either for spectating or as a gameplay change), we could reuse the same field as the field's function is not ddnet-specific.

The downside is that if Teeworlds were to use this field in any other way, we're incompatible with no way of reconciling. Also, it seems unlikely that Teeworlds will ever implement zoom.

**Comment by Fireball-Teeworlds** [comment]
Like this?

**Comment by Fireball-Teeworlds** [comment]
Updated the description - everything is ready from my side.

@heinrich5991 @def- wdyt?

**Comment by def-** [comment]
I think let's get it in and give mappers/users a few weeks to try this out.
bors r+

**Pull request merged**